### PR TITLE
FileState and corresponding metastore & RPC operations

### PR DESCRIPTION
**Pull request by timmyyao**

### CODE REVIEW

File: smart-server/src/test/java/org/smartdata/server/TestSmartClient.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    MetaStore metaStore = ssm.getMetaStore();
+    String path = "/file1";
+
+    FileState fileState = new NormalFileState(path);
+    metaStore.insertUpdateFileState(fileState);
+
+    SmartClient client = new SmartClient(smartContext.getConf());
+    FileState fileState1 = client.getFileState(path);
+    Assert.assertEquals(fileState, fileState1);
+    return;
```
Comment (qiyuangong) [inline smart-server/src/test/java/org/smartdata/server/TestSmartClient.java]: Remove return;

File: smart-common/src/main/java/org/smartdata/model/FileState.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    }
+    FileState fileState = (FileState) o;
+    return path.equals(fileState.getPath())
+        && fileType.equals(fileState.getFileType())
+        && fileStage.equals(fileState.getFileStage());
+  }
+
+  @Override
+  public int hashCode() {
+    return path.hashCode();
```
Comment (littlezhou) [inline smart-common/src/main/java/org/smartdata/model/FileState.java]: Value of hashCode() equals while the two objects are not equal.

File: smart-hadoop-support/smart-hadoop-client/src/main/java/org/apache/hadoop/hdfs/SmartInputStream.java [hunk 3]
```diff
@@ -39,12 +43,29 @@
...
+      throw new RuntimeException("Hadoop version 0.x and 1.x are not supported");
+    }
+    Integer second = Integer.parseInt(parts[1]);
+    if (first == 2 && second <= 6) {
+      return HADOOP_26_VERSION;
+    } else {
+      return HADOOP_27_VERSION;
+    }
+  }
+
```
Comment (timmyyao) [inline smart-hadoop-support/smart-hadoop-client/src/main/java/org/apache/hadoop/hdfs/SmartInputStream.java]: This is a temp realization for Hadoop version check. Need refactor in a more decent way.

File: smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+  public void insertUpate(FileState fileState) {
+    JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
+    String sql = "REPLACE INTO file_state (path, type, stage) VALUES (?,?,?);";
+    jdbcTemplate.update(sql, fileState.getPath(), fileState.getFileType().getValue(),
+        fileState.getFileStage().getValue());
+  }
+
+  public FileState getByPath(String path) {
+    JdbcTemplate jdbcTemplate = new JdbcTemplate(dataSource);
+    return jdbcTemplate.queryForObject("SELECT * FROM file_state WHERE path = ?",
```
Comment (qiyuangong) [inline smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java]: "SELECT * FROM file_state WHERE path = ?". Make "file_state" a property ([ActionDao](https://github.com/Intel-bigdata/SSM/blob/trunk/smart-metastore/src/main/java/org/smartdata/metastore/dao/ActionDao.java#L40)).

File: smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        new Object[]{path}, new FileStateRowMapper());
+  }
```
Comment (qiyuangong) [inline smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java]: Should we add a prefix search method ? For example, find all abc* with abc.

File: smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java (hunk 4 above)
Comment (gnillor) [inline smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java]: Do we need to add row lock like "SELECT * FROM file_state WHERE path = ? FOR UPDATE".

File: smart-metastore/src/main/java/org/smartdata/metastore/utils/MetaStoreUtils.java [hunk 6]
```diff
@@ -338,7 +340,12 @@ public static void initializeDataBase(
               + " dest varchar(4096) NOT NULL,\n"
               + " period bigint(20) NOT NULL\n"
               + ") ;",
-          "CREATE INDEX backup_file_rid_idx ON backup_file (rid);"
+          "CREATE INDEX backup_file_rid_idx ON backup_file (rid);",
+          "CREATE TABLE file_state (\n"
+              + " path varchar(1000) PRIMARY KEY,\n"
```
Comment (gnillor) [inline smart-metastore/src/main/java/org/smartdata/metastore/utils/MetaStoreUtils.java]: Do the file_state will auto sync with the file_info? Like after add a file in hdfs, the file_state will add a row with this file, whose type is normal.

File: smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java (hunk 4 above)
Comment (timmyyao) [reply smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java]: OK. It is done now.

File: smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java (hunk 5 above)
Comment (timmyyao) [reply smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java]: Shall we leave this later when it is indeed needed?

File: smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java (hunk 4 above)
Comment (timmyyao) [reply smart-metastore/src/main/java/org/smartdata/metastore/dao/FileStateDao.java]: Asssign this task to @gnillor and realize it later.

File: smart-metastore/src/main/java/org/smartdata/metastore/utils/MetaStoreUtils.java (hunk 6 above)
Comment (timmyyao) [reply smart-metastore/src/main/java/org/smartdata/metastore/utils/MetaStoreUtils.java]: No, file_state and file_info updated separately.

**Pull request merged**